handler/rfc8628/strategy_hmacsha.go

```diff
@@ -0,0 +1,242 @@
+	Config           interface {
+		fosite.DeviceProvider
+		fosite.DeviceAndUserCodeLifespanProvider
+	}
+}
+
+var _ RFC8628CodeStrategy = (*DefaultDeviceStrategy)(nil)
+
+// GenerateUserCode generates a user_code
+func (h *DefaultDeviceStrategy) GenerateUserCode(ctx context.Context) (string, string, error) {
+	seq, err := randx.RuneSequence(8, []rune(randx.AlphaUpper))
+	if err != nil {
+		return "", "", err
+	}
+	userCode := string(seq)
+	signUserCode, signErr := h.UserCodeSignature(ctx, userCode)
+	if signErr != nil {
+		return "", "", err
+	}
+	return userCode, signUserCode, nil
+}
+
+// UserCodeSignature generates a user_code signature
+func (h *DefaultDeviceStrategy) UserCodeSignature(ctx context.Context, token string) (string, error) {
+	return h.Enigma.GenerateHMACForString(ctx, token)
+}
+
+// ValidateUserCode validates a user_code
+func (h *DefaultDeviceStrategy) ValidateUserCode(ctx context.Context, r fosite.Requester, code string) error {
+	exp := r.GetSession().GetExpiresAt(fosite.UserCode)
+	if exp.IsZero() && r.GetRequestedAt().Add(h.Config.GetDeviceAndUserCodeLifespan(ctx)).Before(time.Now().UTC()) {
+		return errorsx.WithStack(fosite.ErrDeviceExpiredToken.WithHintf("User code expired at '%s'.", r.GetRequestedAt().Add(h.Config.GetDeviceAndUserCodeLifespan(ctx))))
+	}
+	if !exp.IsZero() && exp.Before(time.Now().UTC()) {
+		return errorsx.WithStack(fosite.ErrDeviceExpiredToken.WithHintf("User code expired at '%s'.", exp))
+	}
+	return nil
+}
+
+// GenerateDeviceCode generates a device_code
+func (h *DefaultDeviceStrategy) GenerateDeviceCode(ctx context.Context) (string, string, error) {
+	token, sig, err := h.Enigma.Generate(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	return "ory_dc_" + token, sig, nil
+}
+
+// DeviceCodeSignature generates a device_code signature
+func (h *DefaultDeviceStrategy) DeviceCodeSignature(ctx context.Context, token string) (string, error) {
+	return h.Enigma.Signature(token), nil
+}
+
+// ValidateDeviceCode validates a device_code
+func (h *DefaultDeviceStrategy) ValidateDeviceCode(ctx context.Context, r fosite.Requester, code string) error {
+	exp := r.GetSession().GetExpiresAt(fosite.DeviceCode)
+	if exp.IsZero() && r.GetRequestedAt().Add(h.Config.GetDeviceAndUserCodeLifespan(ctx)).Before(time.Now().UTC()) {
+		return errorsx.WithStack(fosite.ErrDeviceExpiredToken.WithHintf("Device code expired at '%s'.", r.GetRequestedAt().Add(h.Config.GetDeviceAndUserCodeLifespan(ctx))))
+	}
+
+	if !exp.IsZero() && exp.Before(time.Now().UTC()) {
+		return errorsx.WithStack(fosite.ErrDeviceExpiredToken.WithHintf("Device code expired at '%s'.", exp))
+	}
+
+	return h.Enigma.Validate(ctx, strings.TrimPrefix(code, "ory_dc_"))
+}
+
+// ShouldRateLimit is used to decide whether a request should be rate-limited
+func (h *DefaultDeviceStrategy) ShouldRateLimit(context context.Context, code string) (bool, error) {
+	key := code + "_limiter"
+
+	keyBytes := []byte(key)
+	object, err := h.RateLimiterCache.Get(keyBytes)
```

Comment:
This won't work in real world scenarios. The cache is in memory which means:

1. It won't work if more than one instance (which is almost always the case) of the library/hydra running
2. It won't persist across restarts

In my view, rate limiting should be the concern of the firewall/ingress/reverse proxy and I think we could simplify this PR a bit by removing this part or making it optional and/or more generic

Reply:
Agreed on all your points. The reason we included this change is to be spec compliant and the reason we didn't implement a more sophisticated solution (which would probably require an external caching system) is that rate limiting should be handled by the ingress/firewall. But looking back at it, it is probably better to leave it unhandled, instead of having partly correct logic. 

It is probably better to remove this and leave it as future work.